File: python/graph_search/jps.py
'''
@file: jps.py
@breif: Jump Point Search motion planning
@author: Winter
@update: 2023.1.14
'''
import os, sys
import heapq

sys.path.append(os.path.abspath(os.path.join(__file__, "../../")))

from .a_star import AStar
from utils import Env, Node

class JPS(AStar):
    '''
    Class for JPS motion planning.

    Parameters
    ----------
    start: tuple
        start point coordinate
    goal: tuple
        goal point coordinate
    env: Env
        environment
    heuristic_type: str
        heuristic function type, default is euclidean

    Examples
    ----------
    >>> from utils import Grid
    >>> from graph_search import JPS
    >>> start = (5, 5)
    >>> goal = (45, 25)
    >>> env = Grid(51, 31)
    >>> planner = JPS(start, goal, env)
    >>> planner.run()
    '''
    def __init__(self, start: tuple, goal: tuple, env: Env, heuristic_type: str = "euclidean") -> None:
        super().__init__(start, goal, env, heuristic_type)
    
    def __str__(self) -> str:
        return "Jump Point Search(JPS)"

    def plan(self):
        '''
        JPS motion plan function.
        [1] Online Graph Pruning for Pathfinding On Grid Maps

        Return
        ----------
        path: list
            planning path
        expand: list
            all nodes that planner has searched
        '''
        # OPEN set with priority and CLOSED set
        OPEN = []
        heapq.heappush(OPEN, self.start)
        CLOSED = []

        while OPEN:
            node = heapq.heappop(OPEN)

            # exists in CLOSED set
            if node in CLOSED:
                continue

            # goal found
            if node == self.goal:
                CLOSED.append(node)
                return self.extractPath(CLOSED), CLOSED

            jp_list = []
            for motion in self.motions:
                jp = self.jump(node, motion)
                # exists and not in CLOSED set
                if jp and jp not in CLOSED:
                    jp.parent = node.current
                    jp.h = self.h(jp, self.goal)
                    jp_list.append(jp)

            for jp in jp_list:
                # update OPEN set
                heapq.heappush(OPEN, jp)

                # goal found
                if jp == self.goal:
                    break
            
            CLOSED.append(node)
        return [], []

    def jump(self, node: Node, motion: Node):
        '''
        Jumping search recursively.

        Parameters
        ----------
        node: Node
            current node
        motion: Node
            the motion that current node executes

        Return
        ----------
        jump_point: Node
            jump point or None if searching fails
        '''
        # explore a new node
        new_node = node + motion
        new_node.parent = node.current
        new_node.h = self.h(new_node, self.goal)

        # hit the obstacle
        if new_node.current in self.obstacles:
            return None

        # goal found
        if new_node == self.goal:
            return new_node

        # diagonal
        if motion.current[0] and motion.current[1]:
            # if exists jump point at horizontal or vertical
            x_dir = Node((motion.current[0], 0), None, 1, None)
            y_dir = Node((0, motion.current[1]), None, 1, None)
            if self.jump(new_node, x_dir) or self.jump(new_node, y_dir):
                return new_node
            
        # if exists forced neighbor
        if self.detectForceNeighbor(new_node, motion):
            return new_node
        else:
            return self.jump(new_node, motion)

    def detectForceNeighbor(self, node, motion):
        '''
        Detect forced neighbor of node.

        Parameters
        ----------
        node: Node
            current node
        motion: Node
            the motion that current node executes

        Return
        ----------
        flag: bool
            True if current node has forced neighbor else Flase
        '''
        x, y = node.current
        x_dir, y_dir = motion.current

        # horizontal
        if x_dir and not y_dir:
            if (x, y + 1) in self.obstacles and \
                (x + x_dir, y + 1) not in self.obstacles:
                return True
            if (x, y - 1) in self.obstacles and \
                (x + x_dir, y - 1) not in self.obstacles:
                return True
        
        # vertical
        if not x_dir and y_dir:
            if (x + 1, y) in self.obstacles and \
                (x + 1, y + y_dir) not in self.obstacles:
                return True
            if (x - 1, y) in self.obstacles and \
                (x - 1, y + y_dir) not in self.obstacles:
                return True
        
        # diagonal
        if x_dir and y_dir:
            if (x - x_dir, y) in self.obstacles and \
                (x - x_dir, y + y_dir) not in self.obstacles:
                return True
            if (x, y - y_dir) in self.obstacles and \
                (x + x_dir, y - y_dir) not in self.obstacles:
                return True
        
        return False

File: matlab/graph_search/jps.m
function [path, flag, cost, expand] = jps(map, start, goal)
%%
% @file: jps.m
% @breif: Jump Point Search motion planning
% @author: Winter
% @update: 2023.1.29

%{ 
    ============ OPEN and CLOSED==============
    [x, y, g, h, px, py]
    =======================================
%}

%%
% initialize
expand= [];                                                   % expand zone
OPEN  = [start, 0, h(start, goal), start];      % OPEN set with priority 
CLOSED = [];                                                % CLOSED set
flag  = false;
neighbor  = [-1, 1, 1.414;...  
                0, 1, 1;...
                1, 1, 1.414;...
              -1, 0, 1;...
                1, 0, 1;...
              -1, -1, 1.414;...
                0, -1, 1;...
                1, -1, 1.414];
neighbor_num = length(neighbor(:, 1));
cost = 0;

while ~isempty(OPEN(:, 1))
    % pop
    f = OPEN(:, 3) + OPEN(:, 4);
    [~, index] = min(f);
    cur_node = OPEN(index, :);
    OPEN(index, :) = []; 
    
    % exists in CLOSED set
    if loc_list(cur_node, CLOSED, [1, 2])
        continue
    end
    
    % goal found
    if cur_node(1) == goal(1) && cur_node(2) == goal(2)
        CLOSED = [cur_node; CLOSED];
        flag = true;
        cost = cur_node(3);
        break
    end
    
    jp_list = [];
    for i=1:neighbor_num
        [jp, flag] = jump(cur_node, neighbor(i, :), goal, map);
        % exists and not in CLOSED set
        if flag && ~loc_list(jp, CLOSED, [1, 2])
            jp(5:6) = cur_node(1:2);
            jp(4) = h(jp(1:2), goal);
            jp_list = [jp; jp_list];
        end
    end
        
    for j = 1:size(jp_list, 1)
        jp = jp_list(j, :);
        % update OPEN set
        OPEN = [OPEN; jp];
        expand = [expand; jp];

        % goal found
        if jp(1) == goal(1) && jp(2) == goal(2)
            break
        end
    end
    
    CLOSED = [cur_node; CLOSED];
end

% extract path
path = extract_path(CLOSED, start);
end

%%
function h_val = h(cur_node, goal)
% @breif: heuristic function(Manhattan distance)
    h_val = abs(cur_node(1) - goal(1)) + abs(cur_node(2) - goal(2));
end

function index = loc_list(node, list, range)
% @breif: locate the node in given list
    num = size(list);
    index = 0;
    if ~num(1)
        return
    else  
        for i=1:num(1)
            if isequal(node(range), list(i, range))
                index = i;
                return;
            end
        end
    end
end

function path = extract_path(close, start)
% @breif: Extract the path based on the CLOSED set.
    path  = [];               
    closeNum = length(close(:, 1));
    index = 1;
    while 1
        path = [path; close(index, 1:2)];
        if isequal(close(index, 1:2), start)   
            break;
        end
        for i=1:closeNum
            if isequal(close(i, 1:2), close(index, 5:6))
                index = i;
                break;
            end
        end
    end
end

function [new_node, flag] = jump(cur_node, motion, goal, map)
    flag = false;

    % explore a new node
    new_node = [cur_node(1) + motion(1), ...
                      cur_node(2) + motion(2), ...
                      cur_node(3) + motion(3), ...
                      0, cur_node(1), cur_node(2)
                      ];
    new_node(4) = h(new_node(1:2), goal);

    % obstacle
    if new_node(1) <= 0 || new_node(2) <= 0 || map(new_node(1), new_node(2)) == 2
        return
    end

    % goal found
    if new_node(1) == goal(1) && new_node(2) == goal(2)
        flag = true;
        return
    end

    % diagonal
    if motion(1) && motion(2)
        % if exists jump point at horizontal or vertical
        x_dir = [motion(1), 0, 1];
        y_dir = [0, motion(2), 1];
        [~, flag_x] = jump(new_node, x_dir, goal, map);
        [~, flag_y] = jump(new_node, y_dir, goal, map);
        if  flag_x || flag_y 
            flag = true;
            return
        end
    end
            
    % if exists forced neighbor
    if detect_force(new_node, motion, map)
        flag = true;
        return
    else
        [new_node, flag] = jump(new_node, motion, goal, map);
        return
    end
end

function flag = detect_force(cur_node, motion, map)
    flag = true;
    x = cur_node(1);
    y = cur_node(2);
    x_dir = motion(1);
    y_dir = motion(2);

    % horizontal
    if x_dir && ~y_dir
        if map(x, y + 1) == 2 && map(x + x_dir, y + 1) ~= 2
            return
        end
        if map(x, y - 1) == 2 && map(x + x_dir, y - 1) ~= 2
            return
        end
    end
    
    % vertical
    if ~x_dir && y_dir
        if map(x + 1, y) == 2 && map(x + 1, y + y_dir) ~= 2
            return
        end
        if  map(x - 1, y) == 2 && map(x - 1, y + y_dir) ~= 2
            return
        end
    end

    % diagonal
    if x_dir && y_dir
        if map(x - x_dir, y) == 2 && map(x - x_dir, y + y_dir) ~= 2
            return
        end
        if map(x, y - y_dir) == 2 && map(x + x_dir, y - y_dir) ~= 2
            return
        end
    end

    flag = false;
    return
end